File: 2015/day 7/day7.cpp
#include <iostream>
#include <unordered_map>
#include <sstream>
#include <functional>
#include <string>
#include <vector>

unsigned short calculate_wire(const std::string& wire, 
                              std::unordered_map<std::string, std::function<unsigned short()>>& circuit, 
                              std::unordered_map<std::string, unsigned short>& cache) {
    if (isdigit(wire[0])) {
        return static_cast<unsigned short>(std::stoi(wire));
    }

    if (cache.find(wire) == cache.end()) {
        cache[wire] = circuit[wire]();
    }

    return cache[wire];
}

int main() {
    std::unordered_map<std::string, std::function<unsigned short()>> circuit;
    std::unordered_map<std::string, unsigned short> cache;
    std::string line;

    // Replace with actual input
    std::vector<std::string> instructions = {
        "123 -> x",
        "456 -> y",
        "x AND y -> d",
        "x OR y -> e",
        "x LSHIFT 2 -> f",
        "y RSHIFT 2 -> g",
        "NOT x -> h",
        "NOT y -> i"
    };

    for (const auto& instruction : instructions) {
        std::istringstream iss(instruction);
        std::vector<std::string> tokens{std::istream_iterator<std::string>{iss}, 
                                        std::istream_iterator<std::string>{}};

        if (tokens[1] == "->") {
            circuit[tokens[2]] = [tokens, &circuit, &cache] {
                return calculate_wire(tokens[0], circuit, cache);
            };
        } else if (tokens[1] == "AND") {
            circuit[tokens[4]] = [tokens, &circuit, &cache] {
                return calculate_wire(tokens[0], circuit, cache) & calculate_wire(tokens[2], circuit, cache);
            };
        } // Add other operations (OR, LSHIFT, RSHIFT, NOT) here
    }

    // Calculate the values for each wire
    for (const auto& wire : circuit) {
        std::cout << wire.first << ": " << calculate_wire(wire.first, circuit, cache) << std::endl;
    }

    return 0;
}

File: 2015/day 7/day7.py
def get_value(wire, circuit, cache):
    if wire.isdigit():
        return int(wire)
    if wire not in cache:
        operation = circuit[wire]
        if "AND" in operation:
            cache[wire] = get_value(operation[0], circuit, cache) & get_value(operation[2], circuit, cache)
        elif "OR" in operation:
            cache[wire] = get_value(operation[0], circuit, cache) | get_value(operation[2], circuit, cache)
        elif "LSHIFT" in operation:
            cache[wire] = get_value(operation[0], circuit, cache) << int(operation[2])
        elif "RSHIFT" in operation:
            cache[wire] = get_value(operation[0], circuit, cache) >> int(operation[2])
        elif "NOT" in operation:
            cache[wire] = ~get_value(operation[1], circuit, cache) & 0xFFFF
        else:
            cache[wire] = get_value(operation[0], circuit, cache)
    return cache[wire]

def main():
    circuit = {}
    cache = {}
    
    f = open("input.txt", "r")
    input = f.read()
    lines = input.split('\n')

    for line in lines:
        parts = line.split()
        if "->" in line:
            circuit[parts[-1]] = parts[:-2]

    # Calculate the values for each wire
    for wire in circuit:
        print(f"{wire}: {get_value(wire, circuit, cache)}")

if __name__ == "__main__":
    main()

